server/schedulers/adjacent_region.go

```diff
@@ -0,0 +1,307 @@
+	}
+	base := newBaseScheduler(limiter)
+	s := &balanceAdjacentRegionScheduler{
+		baseScheduler: base,
+		opt:           opt,
+		selector:      schedule.NewRandomSelector(filters),
+		leaderLimit:   defaultAdjacentLeaderLimit,
+		peerLimit:     defaultAdjacentPeerLimit,
+		lastKey:       []byte(""),
+	}
+	if len(args) == 2 {
+		s.leaderLimit = args[0]
+		s.peerLimit = args[1]
+	}
+	return s
+}
+
+func (l *balanceAdjacentRegionScheduler) GetName() string {
+	return "balance-adjacent-region-scheduler"
+}
+
+func (l *balanceAdjacentRegionScheduler) GetType() string {
+	return "adjacent-region"
+}
+
+func (l *balanceAdjacentRegionScheduler) GetMinInterval() time.Duration {
+	return minAdjacentSchedulerInterval
+}
+
+func (l *balanceAdjacentRegionScheduler) GetNextInterval(interval time.Duration) time.Duration {
+	return intervalGrow(interval, maxAdjacentSchedulerInterval, linearGrowth)
+}
+
+func (l *balanceAdjacentRegionScheduler) IsScheduleAllowed() bool {
+	return l.allowBalanceLeader() || l.allowBalanceLeader()
+}
+
+func (l *balanceAdjacentRegionScheduler) allowBalanceLeader() bool {
+	return l.limiter.OperatorCount(core.AdjacentLeaderKind) < l.leaderLimit
+}
+
+func (l *balanceAdjacentRegionScheduler) allowBalancePeer() bool {
+	return l.limiter.OperatorCount(core.AdjacentPeerKind) < l.peerLimit
+}
+
+func (l *balanceAdjacentRegionScheduler) Schedule(cluster schedule.Cluster) *schedule.Operator {
+	if l.cacheRegions == nil {
+		l.cacheRegions = &adjacentState{
+			assignedStoreIds: make([]uint64, 0, len(cluster.GetStores())),
+			regions:          make([]*core.RegionInfo, 0, scanLimit),
+			head:             0,
+		}
+	}
+	// we will process cache firstly
+	if l.cacheRegions.len() >= 2 {
+		return l.process(cluster)
+	}
+
+	l.cacheRegions.clear()
+	regions := cluster.ScanRegions(l.lastKey, scanLimit)
+	// scan to the end
+	if len(regions) <= 1 {
+		l.adjacentRegionsCount = 0
+		schedulerStatus.WithLabelValues(l.GetName(), "adjacent_count").Set(float64(l.adjacentRegionsCount))
+		l.lastKey = []byte("")
+		return nil
+	}
+
+	// calculate max adjacentRegions and record to the cache
+	adjacentRegions := make([]*core.RegionInfo, 0, scanLimit)
+	adjacentRegions = append(adjacentRegions, regions[0])
+	maxLen := 0
+	for i, r := range regions[1:] {
+		l.lastKey = r.StartKey
+
+		// append if the region are adjacent
+		lastRegion := adjacentRegions[len(adjacentRegions)-1]
+		if lastRegion.Leader.GetStoreId() == r.Leader.GetStoreId() && bytes.Equal(lastRegion.EndKey, r.StartKey) {
+			adjacentRegions = append(adjacentRegions, r)
+			if i != len(regions)-2 { // not the last element
+				continue
+			}
+		}
+
+		if len(adjacentRegions) == 1 {
+			adjacentRegions[0] = r
+		} else {
+			// got an max length adjacent regions in this range
+			if maxLen < len(adjacentRegions) {
+				l.cacheRegions.clear()
+				maxLen = len(adjacentRegions)
+				l.cacheRegions.regions = append(l.cacheRegions.regions, adjacentRegions...)
+				adjacentRegions = adjacentRegions[:0]
```

Comment:
This seems not right, for example, if `maxLen = 0, len(adjacentRegions) = 2`, then `adjacentRegions` will be clear, but what if the next region is also adjacent?
BTW, the flow in this for loop seems not straight.

Reply:
if next region is also adjacent, the`len(adjacentRegions)` will be 3 not 2. I mean it will not execute to here because of `continue` in L174.